### ISSUE TITLE
fine tuning on several speakers

### ISSUE THREAD

**Issue by omarkshoaib** [edwko/OuteTTS#83]
I already finetuned the model on 500 samples of the same speaker and it worked but still didn't explicitly get the Egyptian dialect  and so i gathered 1500 more samples with there text labels from several speakers.
The question here, should i create for each audio file of the 1500 a speaker profile using the Create_speaker function or should I use group them?

`# TTS Pipeline Setup and Data Preparation
## 1. Setup Environment and Install Dependencies
# Install necessary libraries
%pip install torch torchaudio --extra-index-url https://download.pytorch.org/whl/cu118
%pip install polars datasets transformers accelerate soundfile numpy tqdm loguru

# Install OuteTTS v1.0 specific dependencies
%pip install outetts
%pip install openai-whisper # For Whisper transcription
%pip install descript-audio-codec # For DAC support

# Optional: Install from git if needed (uncomment if necessary)
# %pip install git+https://github.com/OuteAI/OuteTTS.git@main

# Import base libraries
import os
import torch
from google.colab import drive

# Check GPU availability
print(f"CUDA Available: {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"CUDA Device Name: {torch.cuda.get_device_name(0)}")
    device = "cuda"
else:
    device = "cpu"
print(f"Using device: {device}")

# Mount Google Drive
drive.mount('/content/drive')
# Mount Google Drive
drive.mount('/content/drive', force_remount=True)
## 2. Data Preparation
This section loads the data from Google Drive and processes it using the logic from `prepare_data.py`.
pip install OuteTTS
# import pathlib, csv

# wavs_dir = pathlib.Path("/content/drive/MyDrive/audios_chatgbt")
# meta_path = wavs_dir / "metadata.csv"

# # with meta_path.open("w", encoding="utf‑8", newline="") as fh:
# #     writer = csv.writer(fh, delimiter="|")
# #     for wav in wavs_dir.glob("*.wav"):
# #         txt = wav.with_suffix(".txt")
# #         if txt.exists():
# #             transcript = txt.read_text(encoding="utf‑8").strip()
# #             if transcript:
# #                 writer.writerow([wav.name, transcript])
# # print("✅ wrote", meta_path)

import os
import csv
import polars as pl
import torch
from tqdm.notebook import tqdm # Use notebook version for better display
from loguru import logger
import whisper
from outetts.version.v3.utils import create_speaker
from outetts.version.v3.prompt_processor import PromptProcessor
from outetts.dac.interface import DacInterface

class DataPreparation:
    def __init__(self, wavs_dir, metadata_file, save_dir, save_len=5000, model_tokenizer_path="OuteAI/Llama-OuteTTS-1.0-1B"):
        self.wavs_dir = wavs_dir
        self.metadata_file = metadata_file
        self.save_dir = save_dir
        self.save_len = save_len
        self.data = []
        self.save_id = 0
        self.device = "cuda" if torch.cuda.is_available() else "cpu"
        logger.info(f"DataPreparation using device: {self.device}")
        
        # Initialize components for v3
        self.audio_codec = DacInterface(device=self.device)
        self.whisper_model = whisper.load_model("turbo", device=self.device) # Using 'turbo' model as specified
        self.prompt_processor = PromptProcessor(model_tokenizer_path)

    def create_speaker(self, wav_path, transcript=None):
        """
        Create speaker data using the v3 method with Whisper.
        Transcript is optional since Whisper can transcribe, but ground truth is preferred for training.
        """
        try:
            speaker = create_speaker(
                audio_processor=self.audio_codec,
                audio_path=wav_path,
                whisper_model=self.whisper_model,
                whisper_device=self.device
            )
            if transcript:
                speaker["text"] = transcript.strip()
            return speaker
        except Exception as e:
            logger.error(f"Error creating speaker for {wav_path}: {e}")
            return None

    def process(self):
        if not os.path.exists(self.metadata_file):
            logger.error(f"Metadata file not found: {self.metadata_file}")
            return
        if not os.path.isdir(self.wavs_dir):
            logger.error(f"WAVs directory not found: {self.wavs_dir}")
            return

        with open(self.metadata_file, "r", encoding="utf-8") as f:
            reader = csv.reader(f, delimiter="|")
            rows = [row for row in reader if len(row) >= 2 and row[0].strip() and row[1].strip()]

        total = len(rows)
        logger.info(f"Found {total} valid entries in metadata file.")

        for row in tqdm(rows, total=total, desc="Processing audio files"):
            filename, text = row[0].strip(), row[1].strip()
            if not filename.lower().endswith('.wav'):
                filename += '.wav'
            wav_path = os.path.join(self.wavs_dir, filename)

            if not os.path.exists(wav_path):
                logger.warning(f"WAV file {wav_path} does not exist. Skipping.")
                continue
            try:
                speaker = self.create_speaker(wav_path, text)
                if speaker is None:
                    continue
                prompt = self.prompt_processor.get_training_prompt(speaker)
                self.data.append({"prompt": prompt})
            except Exception as e:
                logger.error(f"Error processing {wav_path}: {e}")

            if len(self.data) >= self.save_len:
                self.save_data()

        if self.data:
            self.save_data()

        logger.info("Data preparation finished.")

    def save_data(self):
        os.makedirs(self.save_dir, exist_ok=True)
        save_path = os.path.join(self.save_dir, f"{self.save_id:06d}.parquet")
        logger.info(f"Saving batch {self.save_id} ({len(self.data)} items) to {save_path}")
        try:
            pl.DataFrame(self.data).write_parquet(save_path)
            self.data = []
            self.save_id += 1
        except Exception as e:
            logger.error(f"Failed to save data to {save_path}: {e}")

# --- Configuration ---
# IMPORTANT: Update these paths to point to your data in Google Drive
DRIVE_BASE_PATH = "/content/drive/MyDrive" # Adjust if your data is elsewhere in Drive
# WAVS_DIR = os.path.join(DRIVE_BASE_PATH, "/content/drive/MyDrive/Address/data/wavs") # <-- Original path from notebook
WAVS_DIR = os.path.join(DRIVE_BASE_PATH, "ArabicDataset/wavs") # <-- UPDATE THIS PATH as per instructions
# METADATA_FILE = os.path.join(DRIVE_BASE_PATH, "/content/drive/MyDrive/Address/data/metadata.csv") # <-- Original path from notebook
METADATA_FILE = os.path.join(DRIVE_BASE_PATH, "ArabicDataset/metadata.csv") # <-- UPDATE THIS PATH as per instructions
SAVE_DIR = "/content/prepared_data" # Save processed data within Colab's temporary storage
SAVE_LEN_PER_FILE = 1000 # Adjust batch size for saving if needed (instructions used 1000)
MODEL_TOKENIZER_PATH = "OuteAI/Llama-OuteTTS-1.0-1B"

# --- Execution ---
logger.info("Starting data preparation...")
dp = DataPreparation(WAVS_DIR, METADATA_FILE, SAVE_DIR, save_len=SAVE_LEN_PER_FILE, model_tokenizer_path=MODEL_TOKENIZER_PATH)
dp.process()
logger.info("Data preparation script finished.")
## 3. Train Model
import gc, torch
# After previous run
# del trainer, model, optimizer       # anything that holds tensors
gc.collect()
torch.cuda.empty_cache()

# # Double‑check
# !pip install nvidia-ml-py3 if needed
# import nvidia_smi
# nvidia_smi.nvmlInit()
# h = nvidia_smi.nvmlDeviceGetHandleByIndex(0)
# print("GPU mem now:", nvidia_smi.nvmlDeviceGetMemoryInfo(h).used // 2**20, "MB")
# nvidia_smi.nvmlShutdown()

# # 1️⃣ Remove the too‑new version
# !pip uninstall -y protobuf

# # 2️⃣ Install a compatible one (3.19–3.20.x keep `builder`)
# !pip install "protobuf>=3.19.6,<3.21"
# # 1 ─ clean out the incompatible version
# !pip uninstall -y protobuf

# # 2 ─ install a compatible one
# !pip install "protobuf>=3.19.6,<3.21" --no-binary=protobuf --upgrade

# # 3 ─ verify
# # python - <<'PY'
# import pkg_resources, sys, subprocess, os, importlib
# print("protobuf version:", pkg_resources.get_distribution("protobuf").version)
# # simple sanity import
# from google.protobuf.internal import builder
# print("builder import ok ✓")
# # PY

import glob, os, torch, polars as pl
from loguru import logger
from datasets import Dataset
from transformers import (
    AutoTokenizer, AutoModelForCausalLM,
    Trainer, TrainingArguments,
    DataCollatorForLanguageModeling,
)

# ----------------------------------------------------------------------
# Helper: load prompts stored as *.parquet
# ----------------------------------------------------------------------
def load_dataset_from_parquet(folder: str) -> Dataset:
    prompts = []
    for fpath in glob.glob(os.path.join(folder, "*.parquet")):
        try:
            df = pl.read_parquet(fpath)
            prompts.extend(df["prompt"].to_list())
        except Exception as exc:
            logger.error(f"[load_dataset] {fpath}: {exc}")
    return Dataset.from_dict({"text": prompts})

# ----------------------------------------------------------------------
# Main routine
# ----------------------------------------------------------------------
def main() -> None:
    model_name = "OuteAI/Llama-OuteTTS-1.0-1B"
    data_dir   = "/content/prepared_data"

    tokenizer = AutoTokenizer.from_pretrained(model_name)

    model = AutoModelForCausalLM.from_pretrained(
        model_name,
        device_map="auto",          # fp32 weights (default)
    )
    model.gradient_checkpointing_enable()  # still on

    raw_ds = load_dataset_from_parquet(data_dir)
    raw_ds = raw_ds.train_test_split(test_size=0.1, seed=42)

    def tokenize_fn(batch):
        return tokenizer(
            batch["text"],
            truncation=True,
            max_length=4096,
            padding="max_length",
        )

    tokenised = raw_ds.map(tokenize_fn, batched=True, remove_columns=["text"])
    collator  = DataCollatorForLanguageModeling(tokenizer, mlm=False)

    train_args = TrainingArguments(
        output_dir="/content/fine_tuned_model_2",
        overwrite_output_dir=True,
        num_train_epochs=11,
        per_device_train_batch_size=1,
        per_device_eval_batch_size=1,
        gradient_accumulation_steps=8,   # effective batch 1 × 8 = 8
        gradient_checkpointing=True,     # redundant with .enable(), safe
        eval_strategy="epoch",
        save_strategy="epoch",
        logging_steps=10,
        learning_rate=5e-5,
        weight_decay=0.01,
        remove_unused_columns=False,
        fp16=torch.cuda.is_available(),
        report_to="none",
        save_total_limit=1,
    )

    trainer = Trainer(
        model=model,
        args=train_args,
        train_dataset=tokenised["train"],
        eval_dataset=tokenised["test"],
        data_collator=collator,
    )

    trainer.train()

    model.save_pretrained("/content/fine_tuned_model_2")
    tokenizer.save_pretrained("/content/fine_tuned_model_2")
    logger.info("✅  Model & tokenizer saved to /content/fine_tuned_model_2")

if __name__ == "__main__":
    main()

## 4. Test Model
import shutil, os

src = "/content/fine_tuned_model_2"
dst = "/content/drive/MyDrive/fine_tuned_model_2"

# if you want to overwrite an existing folder in Drive:
if os.path.exists(dst):
    shutil.rmtree(dst)

shutil.copytree(src, dst)
print(f"✅ Copied {src} to {dst}")

ذا مارك فيلل هو مجتمع سكني فاخر في مدينة المستقبل، القاهرة الجديدة، على مساحة ٢٠ فدان. الفلل مصممة بأحدث المعايير وبتتميز بمساحات واسعة وتشطيبات عالية الجودة.

مميزات الفلل:

تصميمات معمارية فريدة.
مساحات خضراء ومناظر طبيعية خلابة.
خدمات متكاملة زي حمامات السباحة، النوادي، ومناطق للترفيه.
الفلل في ذا مارك بتقدم تجربة معيشية مميزة بتجمع بين الفخامة والراحة في بيئة آمنة وهادئة. لو عايز تعرف تفاصيل أكتر، ممكن تزور الموقع الرسمي لشركة ذا مارك أو تتواصل مع المطور. لو محتاج أي مساعدة إضافية، أنا هنا للمساعدة!

في كمبوند أب تاون كايرو بالمقطم، تتوفر فلل مستقلة بمساحة ٤٨٥ متر مربع. هذه الفلل تتكون من ٤ غرف نوم، ٤ حمامات، وغرفة للخادمة بحمام خاص. تتميز الفلل بتشطيبات عالية الجودة وإطلالات خلابة على المدينة.

السعر الإجمالي للفيلا يبدأ من ١٠٤,٢٧٦,٨٨٨ جنيه مصري، مع دفعة مقدمة بنسبة ٣٠%، والباقي بالتقسيط حتى عام ٢٠٣١.

لمزيد من التفاصيل، يمكنك زيارة موقع Property Finder.
في مدينة العبور، فيه شقق دوبلكس بمساحة 220 متر مربع. مثلاً، في الحي السادس، فيه شقة دوبلكس أول بلكونة بجوار سوبر ماركت جرين، متاحة للبيع. وفي الحي الأول، فيه شقة دوبلكس 220 متر ناصيتين، قريبة من محطة المترو. الأسعار بتختلف حسب الموقع والتشطيب. لو محتاج تفاصيل أكتر عن شقة معينة، قولي.
import outetts
import os

# Use the same output directory as the training step
fine_tuned = "/content/fine_tuned_model_2" 
assert os.path.isdir(fine_tuned), f"Fine-tuned model directory not found: {fine_tuned}"

# Configure the interface using v3 auto_config
cfg = outetts.ModelConfig.auto_config(
    model=outetts.Models.VERSION_1_0_SIZE_1B,
    backend=outetts.Backend.HF,
)
# Point to the fine-tuned model and tokenizer paths
cfg.model_path = fine_tuned
cfg.tokenizer_path = fine_tuned

# Initialize the v3 interface
interface = outetts.Interface(config=cfg)

# Load a speaker (use default English for now, replace if an Arabic speaker profile is available)
# To create an Arabic speaker: speaker = interface.create_speaker("path/to/arabic_audio.wav")
speaker = interface.load_default_speaker("EN-FEMALE-1-NEUTRAL") 

# Configure generation with Arabic text
gen_cfg = outetts.GenerationConfig(
    text="في كمبوند ذا مارك بالتجمع الخامس، تتوفر وحدات سكنية متنوعة للبيع.", # Example Arabic text from instructions
    speaker=speaker,
    generation_type=outetts.GenerationType.CHUNKED,
    sampler_config=outetts.SamplerConfig(
        temperature=0.7,
        top_p=0.9,
        repetition_penalty=1.03,
    ),
)

# Generate the audio
wav = interface.generate(config=gen_cfg)

# Save the output
output_path = "/content/test_output_arabic.wav"
wav.save(output_path)
print(f"✓ Saved {output_path}")

# this was finetuning using sft
# now i will attempt again using lora(peft
)
@`
